test/org/javarosa/core/model/ItemsetBindingTest.java

```diff
@@ -0,0 +1,147 @@
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+
+import static org.javarosa.core.test.AnswerDataMatchers.stringAnswer;
+import static org.javarosa.core.test.SelectChoiceMatchers.choice;
+
+public class ItemsetBindingTest {
+    private Scenario scenario;
+
+    @Before
+    public void setUp() {
+        scenario = Scenario.init("three-level-cascading-select.xml");
+    }
+
+    @Test
+    public void dependentLevelsInBlankInstance_ShouldHaveNoChoices() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+    }
+
+    @Test
+    public void selectingValueAtLevel1_ShouldFilterChoicesAtLevel2() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+
+        scenario.answer("/data/level1", "b");
+
+        assertThat(scenario.choicesOf("/data/level2"), containsInAnyOrder(
+            choice("ba"),
+            choice("bb"),
+            choice("bc")));
+    }
+
+    @Test
+    public void selectingValuesAtLevels1And2_ShouldFilterChoicesAtLevel3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "b");
+        scenario.answer("/data/level2", "ba");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("baa"),
+            choice("bab")));
+    }
+
+    @Test
+    public void clearingValueAtLevel2_ShouldClearChoicesAtLevel3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "a");
+        scenario.answer("/data/level2", "aa");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("aaa"),
+            choice("aab")));
+        scenario.answer("/data/level2", "");
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+    }
+
+    @Test
+    public void clearingValueAtLevel1_ShouldClearChoicesAtLevels2And3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "a");
+        scenario.answer("/data/level2", "aa");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("aaa"),
+            choice("aab")));
+
+        scenario.answer("/data/level1", "");
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        // this next assertion is only true because the one before called populateDynamic choices
```

Comment:
This leads me to think that we should change how the Scenario works. Instead of surgically writing a value into the main instance, I think it should follow the same workflow as Collect, which triggers all the side-effects this test relies upon.

Reply:
`Scenario` works great! It writes the value such that the form is re-evaluated with the new value. I mean that all of the expressions (in constraints, calculates, etc) with values that might have changed are re-evaluated. In code, that means `IDag.triggerTriggerables` is called.

The problem is with the implementation -- `populateDynamicChoices` should be called if and only if values that are part of the expression that defines the itemset are changed. For the first level select of the first cascade in the test form, the expression would be `instance('level2')/root/item[level1 = /data/level1]` and only a change to `/data/level1` should trigger a re-evaluation of the itemset expression.

Without the DAG-based optimization, I'd expect all expressions to be re-evaluated every time any value in the form instance changes. That's not what happens in the case of these choice lists and that's good because it would slow answering any question down a lot. Instead it's this weird in-between -- the itemset declaration is not part of the normal form evaluation path and it's only when the choice list is requested that it's recomputed. That's much better than every time a value is saved but not as good as only when a dependent value is evaluated.

I'm evidently having trouble explaining this. Please ask more questions if it's still not clear.

Reply:
I think you nailed the explanation :)

I didn't consider that a DAG-based optimization could perform worse. Maybe combining both is what makes more sense? i.e. rebuilding choice lists when requesting it, only when a dependent value has changed. QuestionDefs could save a snapshot of their last set of choices for fast access to them.

Reply:
> I didn't consider that a DAG-based optimization could perform worse.

I hadn't either, hah. I see what you mean -- it's possible that the choice list won't be requested and in that case triggering the update on dependent value change would mean a lot of work being done for nothing. I think that's not going to be a very common scenario because generally if the choices change, the question has to be answered again.

That said, I like the way that you are thinking. I was thinking of treating the expression that gets evaluated and putting it in the DAG with its dependent values triggering a re-evaluation but perhaps there's a way for `ItemsetBinding` to know its dependents and find out whether any have changed when choices are requested. I'm not sure that's simpler but it's something to explore.

> QuestionDefs could save a snapshot of their last set of choices for fast access to them.

`ItemsetBinding` does that, right? It's just that the choices are currently always cleared?